### PR TITLE
gh-115754: Use Py_GetConstant(Py_CONSTANT_EMPTY_BYTES)

### PR DESCRIPTION
**Pull request by vstinner**
Replace PyBytes_FromStringAndSize(NULL, 0)
with Py_GetConstant(Py_CONSTANT_EMPTY_BYTES).

* Issue: gh-115754

### CODE REVIEW

**Review by erlend-aasland — APPROVED** [review]

**Review by serhiy-storchaka — COMMENTED** [review]
Please wait.

**Review by serhiy-storchaka — COMMENTED** [review]
I am not comfortable with this change.

1. There is nothing wrong in using `PyBytes_FromStringAndSize(NULL, 0)`. It works, it worked, and it will always work. This is not inefficient or private API.
2. `PyBytes_FromStringAndSize()` returns a new reference, while `Py_GetConstant()` returns a borrowed reference. Even if an empty bytes object is immutable, there is a value of always using balanced incref/decref. It makes the intention clearer and the code more error-proof. When there are several exit points and some return a new reference and other return borrowed reference, you can immediately say that there is a bug. If it is not a bug, you will waste time and mental efforts in vain on analyzing the code.
3. This creates a tiny obstacle for future backportings.

In summary, this is a cosmetic change which makes the code more difficult to maintain without significant gain.

**Comment by vstinner** [comment]
> PyBytes_FromStringAndSize() returns a new reference, while Py_GetConstant() returns a borrowed reference.

Py_GetConstant() returns a new reference: https://docs.python.org/dev/c-api/object.html#c.Py_GetConstant

**Comment by serhiy-storchaka** [comment]
Oh, nice. That eliminates most of my objections. There is still an issue with reasoning such change. This look like change for the sake of change.

**Comment by vstinner** [comment]
I would like to get rid of `PyBytes_FromStringAndSize(NULL, size)` calls which create an [incomplete bytes object](https://github.com/capi-workgroup/api-evolution/issues/36), but `PyBytes_FromStringAndSize(NULL, 0)` doesn't have this issue.

`Py_GetConstant(Py_CONSTANT_EMPTY_BYTES)` is simpler and a little bit faster than `PyBytes_FromStringAndSize(NULL, 0)`.

In general, I would prefer to make `Py_GetConstant(Py_CONSTANT_EMPTY_BYTES)` the reference function to get an empty byte string.

**Comment by vstinner** [comment]
See also https://discuss.python.org/t/what-do-you-think-of-deprecating-c-apis-to-modify-immutable-strings/67057

**Comment by vstinner** [comment]
I failed to convince @serhiy-storchaka, so I just close my PR.
